File: apps/mobile/ios/Sources/FlagshipCore/ProvisionProgress.swift
import Foundation

/// Provisioning progress model for the "your server is being installed"
/// UI. Swift mirror of packages/protocol/src/provisionProgress.ts — the
/// fraction, the four-group labels, and the per-step states must match
/// the webapp + Android renderers byte-for-byte (validated by the shared
/// phase ladder + the conformance tests).
public enum ProvisionProgress {

    /// The fine-grained ladder, in order, EXCLUDING the terminal
    /// `failed` phase. Mirror of PROVISION_PHASES minus `failed`.
    public static let ladder: [String] = [
        "boot",
        "cloned",
        "deps",
        "built",
        "identity",
        "registered",
        "tunnel-online",
        "acme-order",
        "dns01-publish-attempt",
        "dns01-publish-ok",
        "dns01-propagation-wait",
        "tlsalpn-served",
        "acme-validating",
        "cert-issued",
        "ready",
    ]

    /// Human title per fine-grained phase. Lockstep with the protocol's
    /// PROVISION_PHASE_TITLES + the control-plane push fan-out titles, so
    /// the in-app step copy matches the push the user just tapped.
    public static let phaseTitles: [String: String] = [
        "boot": "Server booting",
        "cloned": "Code cloned",
        "deps": "Installing dependencies",
        "built": "Build complete",
        "identity": "Identity generated",
        "registered": "Registered with Flagship",
        "tunnel-online": "Tunnel online",
        "acme-order": "Requesting certificate",
        "dns01-publish-attempt": "Publishing DNS challenge",
        "dns01-publish-ok": "DNS challenge published",
        "dns01-propagation-wait": "Waiting for DNS",
        "tlsalpn-served": "Serving TLS challenge",
        "acme-validating": "Validating certificate",
        "cert-issued": "TLS certificate issued",
        "ready": "Server is live",
        "failed": "Provisioning failed",
    ]

    public enum StepKey: String, Sendable, Equatable {
        case booting, registering, securing, ready
    }

    public struct StepGroup: Sendable, Equatable {
        public let key: StepKey
        public let label: String
        public let phases: [String]
    }

    /// The four user-facing groups, in order.
    public static let stepGroups: [StepGroup] = [
        StepGroup(key: .booting, label: "Booting",
                  phases: ["boot", "cloned", "deps", "built", "identity"]),
        StepGroup(key: .registering, label: "Registering",
                  phases: ["registered", "tunnel-online"]),
        StepGroup(key: .securing, label: "Securing (TLS certificate)",
                  phases: [
                    "acme-order",
                    "dns01-publish-attempt",
                    "dns01-publish-ok",
                    "dns01-propagation-wait",
                    "tlsalpn-served",
                    "acme-validating",
                    "cert-issued",
                  ]),
        StepGroup(key: .ready, label: "Ready", phases: ["ready"]),
    ]

    public enum StepState: String, Sendable, Equatable {
        case done, active, pending, failed
    }

    public struct StepView: Sendable, Equatable {
        public let key: StepKey
        public let label: String
        public let state: StepState
        /// Fine-grained phase title for the ACTIVE / FAILED group; nil
        /// otherwise.
        public let detail: String?
        public init(key: StepKey, label: String, state: StepState, detail: String?) {
            self.key = key
            self.label = label
            self.state = state
            self.detail = detail
        }
    }

    private static func isLadderPhase(_ p: String) -> Bool {
        ladder.contains(p)
    }

    /// Map a phase to a 0..1 fraction for a determinate progress bar.
    public static func fraction(_ phase: String?) -> Double {
        guard let phase, !phase.isEmpty else { return 0 }
        if phase == "ready" { return 1 }
        if phase == "failed" { return 0 }
        guard let idx = ladder.firstIndex(of: phase) else { return 0 }
        return Double(idx + 1) / Double(ladder.count)
    }

    private static func groupKey(forPhase phase: String) -> StepKey {
        for g in stepGroups where g.phases.contains(phase) { return g.key }
        return .booting
    }

    /// Project (phase, lastError, prevPhase) into the per-group checklist
    /// every detail page renders. See the protocol module for the rules.
    public static func stepStates(
        phase: String?,
        lastError: String? = nil,
        prevPhase: String? = nil
    ) -> [StepView] {
        if phase == "ready" {
            return stepGroups.map {
                StepView(key: $0.key, label: $0.label, state: .done, detail: nil)
            }
        }

        if phase == "failed" {
            let failedPhase = (prevPhase.map(isLadderPhase) ?? false) ? prevPhase! : "boot"
            let failedGroup = groupKey(forPhase: failedPhase)
            let failedIdx = stepGroups.firstIndex { $0.key == failedGroup } ?? 0
            return stepGroups.enumerated().map { (i, g) in
                if i < failedIdx {
                    return StepView(key: g.key, label: g.label, state: .done, detail: nil)
                }
                if i == failedIdx {
                    let d = (lastError?.isEmpty ?? true) ? phaseTitles["failed"] : lastError
                    return StepView(key: g.key, label: g.label, state: .failed, detail: d)
                }
                return StepView(key: g.key, label: g.label, state: .pending, detail: nil)
            }
        }

        guard let phase, isLadderPhase(phase) else {
            // No checkpoint yet → first group active, no detail.
            return stepGroups.enumerated().map { (i, g) in
                StepView(key: g.key, label: g.label, state: i == 0 ? .active : .pending, detail: nil)
            }
        }

        let activeGroup = groupKey(forPhase: phase)
        let activeIdx = stepGroups.firstIndex { $0.key == activeGroup } ?? 0
        return stepGroups.enumerated().map { (i, g) in
            if i < activeIdx {
                return StepView(key: g.key, label: g.label, state: .done, detail: nil)
            }
            if i == activeIdx {
                return StepView(key: g.key, label: g.label, state: .active, detail: phaseTitles[phase])
            }
            return StepView(key: g.key, label: g.label, state: .pending, detail: nil)
        }
    }

    /// Should the list (Home) render a thin progress bar for this demo
    /// server? True for any pre-`ready` server; false for ready / none /
    /// absent (those render as a normal online / empty row).
    public static func shouldShowProgressBar(phase: String?, status: String?) -> Bool {
        if status == "up", phase == nil || phase == "ready" { return false }
        if phase == "ready" { return false }
        if status == "none" { return false }
        return status == "provisioning" || phase != nil
    }
}

File: packages/protocol/src/provisionProgress.ts
/**
 * Provisioning progress model — the single source of truth that the
 * "your server is being installed" UI on every surface mirrors.
 *
 * The fine-grained `PROVISION_PHASES` ladder (auth.ts) is what the box
 * actually pushes. A user-facing progress bar wants two derived things:
 *
 *   1. a 0..1 fraction so a thin determinate bar can render anywhere, and
 *   2. a small set of NAMED, grouped steps ("Booting", "Registering",
 *      "Securing (TLS certificate)", "Ready") with per-step state so a
 *      detail page can show a checklist and the user can say "my device
 *      is stuck at Securing".
 *
 * Both are derived here so iOS / Android / webapp stay byte-aligned (the
 * grouping + labels are reproduced verbatim on each surface and unit-
 * tested against this module's outputs via the shared phase ladder).
 */

import {
  PROVISION_PHASES,
  type ProvisionPhase,
  isProvisionPhase,
} from "./auth.js";

/** The ordered, non-terminal ladder (everything except `failed`). The
 *  fraction is computed against this so `ready` is exactly 1.0. */
export const PROVISION_LADDER: readonly ProvisionPhase[] =
  PROVISION_PHASES.filter((p) => p !== "failed");

/**
 * Map a phase to a 0..1 fraction for a determinate progress bar.
 *
 *   - `ready`              → 1
 *   - any ladder phase     → (index + 1) / ladderLength   (monotonic)
 *   - `failed`             → the fraction of the step it died ON, when
 *                            known. We don't track which phase preceded
 *                            `failed` here (the row only stores the
 *                            latest phase); callers that DO know the
 *                            pre-failure phase should pass it. With no
 *                            hint we return 0 so a bare `failed` renders
 *                            as "stalled at the start" rather than
 *                            falsely "almost done".
 *   - unknown / null       → 0
 */
export function provisionFraction(
  phase: string | null | undefined,
): number {
  if (!phase) return 0;
  if (phase === "ready") return 1;
  if (phase === "failed") return 0;
  if (!isProvisionPhase(phase)) return 0;
  const idx = PROVISION_LADDER.indexOf(phase);
  if (idx < 0) return 0;
  return (idx + 1) / PROVISION_LADDER.length;
}

/** The four user-facing groups, in order. The keys are stable ids the
 *  UI keys off; `label` is the human copy reproduced on each surface. */
export type ProvisionStepKey = "booting" | "registering" | "securing" | "ready";

export interface ProvisionStepGroup {
  key: ProvisionStepKey;
  label: string;
  /** The ladder phases that roll up into this group, in order. */
  phases: readonly ProvisionPhase[];
}

/** The grouping. `ready` is its own terminal group so the checklist's
 *  final row is the celebratory "Ready" rather than a sub-step. */
export const PROVISION_STEP_GROUPS: readonly ProvisionStepGroup[] = [
  {
    key: "booting",
    label: "Booting",
    phases: ["boot", "cloned", "deps", "built", "identity"],
  },
  {
    key: "registering",
    label: "Registering",
    phases: ["registered", "tunnel-online"],
  },
  {
    key: "securing",
    label: "Securing (TLS certificate)",
    phases: [
      "acme-order",
      "dns01-publish-attempt",
      "dns01-publish-ok",
      "dns01-propagation-wait",
      "tlsalpn-served",
      "acme-validating",
      "cert-issued",
    ],
  },
  {
    key: "ready",
    label: "Ready",
    phases: ["ready"],
  },
];

/** Per-step render state for the detail-page checklist. */
export type ProvisionStepState = "done" | "active" | "pending" | "failed";

export interface ProvisionStepView {
  key: ProvisionStepKey;
  label: string;
  state: ProvisionStepState;
  /** Sub-step copy for the ACTIVE (or FAILED) group — the fine-grained
   *  phase title, e.g. "Waiting for DNS". Null for non-active groups. */
  detail: string | null;
}

/** Human-readable title for each fine-grained phase. Kept in lockstep
 *  with PHASE_TITLES in packages/control-plane/src/provisionEvents.ts
 *  (the push fan-out's titles) so the in-app step copy matches the
 *  push the user just tapped. */
export const PROVISION_PHASE_TITLES: Record<ProvisionPhase, string> = {
  boot: "Server booting",
  cloned: "Code cloned",
  deps: "Installing dependencies",
  built: "Build complete",
  identity: "Identity generated",
  registered: "Registered with Flagship",
  "tunnel-online": "Tunnel online",
  "acme-order": "Requesting certificate",
  "dns01-publish-attempt": "Publishing DNS challenge",
  "dns01-publish-ok": "DNS challenge published",
  "dns01-propagation-wait": "Waiting for DNS",
  "tlsalpn-served": "Serving TLS challenge",
  "acme-validating": "Validating certificate",
  "cert-issued": "TLS certificate issued",
  ready: "Server is live",
  failed: "Provisioning failed",
};

/** Which group does a (non-terminal) phase belong to? */
function groupKeyForPhase(phase: ProvisionPhase): ProvisionStepKey {
  for (const g of PROVISION_STEP_GROUPS) {
    if ((g.phases as readonly string[]).includes(phase)) return g.key;
  }
  return "booting";
}

/**
 * Project a (phase, lastError, prevPhase) tuple into the per-group
 * checklist state every detail page renders.
 *
 *   - phase==='ready'  → all groups done.
 *   - phase==='failed' → the group that owns `prevPhase` (the last
 *     non-terminal phase, when known) is marked `failed` and carries
 *     `lastError`; groups before it are `done`, groups after `pending`.
 *     With no `prevPhase` hint the FIRST group fails (conservative —
 *     a bare `failed` means "stalled early").
 *   - any other phase  → the owning group is `active` (and carries the
 *     fine-grained phase title as `detail`); earlier groups `done`,
 *     later groups `pending`.
 *   - null phase       → first group `active` with no detail (just
 *     started; no checkpoint yet).
 *
 * NOTE: the daemon now RETRIES on ACME failures, so a UI should render
 * `failed` as "retrying — last error: …", not a dead end. This function
 * surfaces the failure state + error; the copy is the surface's job.
 */
export function provisionStepStates(
  phase: string | null | undefined,
  lastError?: string | null,
  prevPhase?: string | null,
): ProvisionStepView[] {
  const groups = PROVISION_STEP_GROUPS;

  if (phase === "ready") {
    return groups.map((g) => ({
      key: g.key,
      label: g.label,
      state: "done" as const,
      detail: null,
    }));
  }

  if (phase === "failed") {
    const failedPhase: ProvisionPhase =
      prevPhase && isProvisionPhase(prevPhase) && prevPhase !== "failed"
        ? prevPhase
        : "boot";
    const failedGroup = groupKeyForPhase(failedPhase);
    const failedIdx = groups.findIndex((g) => g.key === failedGroup);
    return groups.map((g, i) => {
      if (i < failedIdx) {
        return { key: g.key, label: g.label, state: "done" as const, detail: null };
      }
      if (i === failedIdx) {
        return {
          key: g.key,
          label: g.label,
          state: "failed" as const,
          detail: lastError && lastError.length > 0 ? lastError : PROVISION_PHASE_TITLES.failed,
        };
      }
      return { key: g.key, label: g.label, state: "pending" as const, detail: null };
    });
  }

  // No checkpoint yet → first group active, no detail.
  if (!phase || !isProvisionPhase(phase)) {
    return groups.map((g, i) => ({
      key: g.key,
      label: g.label,
      state: i === 0 ? ("active" as const) : ("pending" as const),
      detail: null,
    }));
  }

  const activeGroup = groupKeyForPhase(phase);
  const activeIdx = groups.findIndex((g) => g.key === activeGroup);
  return groups.map((g, i) => {
    if (i < activeIdx) {
      return { key: g.key, label: g.label, state: "done" as const, detail: null };
    }
    if (i === activeIdx) {
      return {
        key: g.key,
        label: g.label,
        state: "active" as const,
        detail: PROVISION_PHASE_TITLES[phase],
      };
    }
    return { key: g.key, label: g.label, state: "pending" as const, detail: null };
  });
}
